Use layered class loaders rather than "shared" class loader hack for class loader isolation

alexarchambault opened the pull request:
Same as https://github.com/com-lihaoyi/mill/pull/5155, but rebased on the latest main. Opening a PR here to more easily check what the CI says, and not do a push-force on an upstream open PR. Not meant to be merged here.

alexarchambault committed:
Set context class loader when running tasks

autofix-troubleshooter commented:
Hi! I'm the <a href="https://autofix.ci"><img alt="autofix logo" src="https://autofix.ci/logo/logo.svg" width="16"><strong>autofix.ci</strong></a> troubleshooter bot.

It looks like you correctly set up a CI job that uses the autofix.ci GitHub Action, but the [**autofix.ci GitHub App**](https://github.com/apps/autofix-ci/) has not been installed for this repository. This means that autofix.ci unfortunately does not have the permissions to fix this pull request. If you are the repository owner, please install the app and then restart the CI workflow! 😃

alexarchambault committed:
Manage coursier env vars via CoursierConfigModule

alexarchambault committed:
RefCountedClassLoaderCache.extraRelease for VerifyError

alexarchambault committed:
Minor rework of default repositories handling

alexarchambault committed:
Re-enable mirror test

alexarchambault committed:
Start Mill with a layered class loader

lefou committed:
Support `ScalaNativeModule.runMain` (#5723)

For each main class, we need an additional linking task. Due to the
fact, that we don't know the name of the main class to run before, we
currently can't cache the result of the linking task for user-provided
main classes.

See discussion for a potential solution:
*
https://github.com/com-lihaoyi/mill/discussions/4595#discussioncomment-14185978
*
https://github.com/com-lihaoyi/mill/discussions/4595#discussioncomment-14195897

Also replaced the directly defined `run` task by implementing `runner():
Runner`, so that users can run native mains in more situations. The
`run` method is now obsolete and is kept to do the same as `super.run()`
to preserve binary compatibility.

Fix https://github.com/com-lihaoyi/mill/issues/5461

Pull request: https://github.com/com-lihaoyi/mill/pull/5723

arturaz committed:
Easier to read formatting in `ExampleTester.scala`.

arturaz committed:
Print logs to console if an exception happens during mill launcher initialization (#5736)

Currently if an exception like
https://github.com/com-lihaoyi/mill/issues/5734 happens while launching
the mill daemon there is no logs to look to. This PR adds in-memory
logging and outputs that in case of failure.

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>

lihaoyi committed:
Avoid using system `java` unless explicitly specified, instead use latest LTS Java version unless explicitly overridden with `system` (#5705)

This would make Mill builds more reproducible, and avoid the JVM-version
confusion that e.g. @makingthematrix encountered recently in the discord
channel.

The basic question is what to do if `//| mill-jvm-version` is not
explicitly set. It seems all options have tradeoffs

1. Not providing any default and forcing the user to provide one
* Annoying for first-time users, which have to manually set a `//|
mill-jvm-version` even if they don't care about it
* Even plenty of advanced users or large projects don't _really_ care
about their JVM version, so asking them to pick one is annoying when we
could just pick something for them and they'll be fine
* If Mill aims to support non-JVM toolchains like Python or JS, then
they don't care about this either and would be happy for some default to
be picked
* We'd need special handling for integration and example tests, and
internal uses like `dist.run` or `dist.installLocal`, where providing an
explicit JVM for every test case would be super annoying. But special
handling would make them diverge from normal build files (although they
already diverge, e.g. not using a bootstrap script and not requiring a
`mill-version` specified)
 
2. Using some hardcoded default even if there is a `java`
environmentally installed
* Can cause confusion since people would expect the `java` they
installed to take effect in the absence of other explicit configuration
* Anyone upgrading from earlier versions of Mill could have their JVM
version change silently, possibly causing published artifacts to
silently require newer JVMs to run and causing downstream bytecode
validation errors later
* If we keep the default version stable, it will inevitable fall out of
date over time. If we keep the default version updated, it might change
the user's JVM version every time they update Mill
* There is no "generally safe" default version to use: libraries would
prefer the lowest possible version to not impose requirements on
downstream users, while applications would prefer the highest possible
version to take advantage of JVM improvements. `com-lihaoyi` libraries
are built with 11, Spring-Boot requires 17, etc.
* The existence of a default version that the user has to notice is
incorrect and then override adds substantial user-facing complexity over
forcing them to pick a version in the first place, since in the latter
case at least the user would be aware of the value and it's ability to
be configured

3. Using the environmental `java` installation as the default, 
* Matches the "default" way most people work with JVMs, but that default
way sucks
* Someone who checks out a codebase on a different machine with some
other machine with a different `java` installation would encounter
failures.
* While users could always resolve this issue by setting an explicit
`//| mill-jvm-version` in their build header, most users likely won't by
default, won't know that they should, and when they hit weird Java
version errors they won't know that this is the solution for it. e.g.
all the `com-lihaoyi` projects never did this, and if we as the experts
aren't doing this ourselves we can't realistically expect others to do
so

We are currently at (3), but of these options I think (2) is probably
the best:

(1) Adds significant increased up-front inconvenience. This can be
mitigated by copy-pasteable instructions on `installation-ide.adoc`, but
then people are likely to end up with the wrong JVM version anyway by
following the default, and so there's not that much value in forcing it
to be visible in the build file if they don't care about it

(2) Has the problem that using the non-system version is surprising, but
has the advantage that people who "don't care" about their JVM version
don't need to do/write/see anything, and those that do care can just set
it. We can mitigate this problem by putting it in the changelog and docs
where people install Mill and configure JVM versions, with instructions
on how to set it

The longer-term convenience of "Mill projects are easy for anyone to
checkout and build" would definitely be very valuable, and can be a
differentiator over other build tools where that is generally not the
case

As the default version, the policy is to choose the latest LTS version
of the JVM, which at this point in time is 21.

This is a breaking change, but it doesn't break binary compatibility, so
it'll likely go into 1.1.0. Users can fall back to the prior behavior
using `mill-jvm-version: system` or `mill-jvm-version: 17` as necessary

lihaoyi committed:
Add integration test for `mill-jvm-version: system` (#5740)

Follow up of https://github.com/com-lihaoyi/mill/pull/5705

lefou committed:
Cleanup of scalalib.test files (#5750)

Fixed package declarations, Scala 2 wildcards, missing empty lines
between import and class.

Pull request: https://github.com/com-lihaoyi/mill/pull/5750

lihaoyi committed:
Rebootstrap on 1.0.4 (#5752)

lefou committed:
Update jsoup to 1.21.2 (in meta-build) (#5753)

Pull request: https://github.com/com-lihaoyi/mill/pull/5753

lefou committed:
Implement support for Scala 3.8 (#5751)

Instead of using a scala-library-2.13.x, since Scala 3.8.0 the used
scala-library has the same version as the Scala version.

Fix https://github.com/com-lihaoyi/mill/issues/5739

Co-authored-by: Natsu Kagami <nki@nkagami.me>

vaslabs committed:
KSP 2 support (#5727)

## Motivation: 

https://github.com/com-lihaoyi/mill/issues/5676

Also, language version 1.9 is deprecated since Kotlin 2.2.0. 

It turns out, that integration wise, it's much easier to run the symbol
processor with KSP 2 via a CLI wrapper that it is for KSP 1 as a kotlin
compiler plugin


## Details

The documentation for the CLI wrapper is
[here](https://github.com/google/ksp/blob/main/docs/ksp2cmdline.md)
albeit a bit outdated. The key component that is missing, is a reference
to the `-libraries` argument, which (as our trial and error shows)
expects the module classpath.

Micronaut had also a tricky bit, failing to resolve java sdk sources.
Passing the `-jdk-home` resolves that.

From there on, Dagger, Hilt and micronaut work without any major
adaptations from the user's api perspective as we made an effort to
consolidate `KSPModule` and `KSP2Module` input parameters

We have also retested the pokemon module outside of this repo which also
works, surprisingly without needing the `-friends` parameter

## Summary

This PR gives mill the capability to build KSP 2 with language version
and api version >= 2.0 , bringing a solution to the deprecated language
version <=1.9 needed by KSP 1.

Also, added an equivalent Micronaut example as the hello-micronaut in
javalib/web, but with full KSP, no annotation processors needed

## Credits

This PR was co-authored by @souvlakias

---------

Co-authored-by: souvlakias <stylianos@vaslabs.io>

lihaoyi committed:
Use `DoubleLock` for mutex on compiler bridge creation to avoid races between `MillDaemonMain` and `ZincWorkerMain` processes (#5761)

Hopefully fixes https://github.com/com-lihaoyi/mill/issues/5743. Will
need to rebootstrap on this to stop the failures in our CI

The previous in-memory lock is no longer sufficient for mutex in this
code path, because now both the main Mill process and any
`ZincWorkerMain` subprocess could both be trying to initialize compiler
bridges in the same `compilerBridge.workspace` folder

lihaoyi committed:
Rebootstrap (#5763)

Pulls in
https://github.com/com-lihaoyi/mill/commit/28c1bd4851e4582ac1e2e6a1562ab35cfd4fcbb7
to try and fix the lint-autofix flakiness

gamlerhart committed:
Fix References in Developer docs (#5764)

- Replace the 'main' with the 'core' module
- Fix reference to the DebugLog

Pull request: https://github.com/com-lihaoyi/mill/pull/5764

vaslabs committed:
Android: Package jnis provided by dependencies (#5760)

This PR resolves an issue found [during
integration](https://github.com/com-lihaoyi/mill/issues/5758) where we
were missing the functionality to include in the apk, native deps
provided by dependencies

<img width="1448" height="637" alt="image"
src="https://github.com/user-attachments/assets/8101b58f-5006-4cac-9fc8-8b277b3007ab"
/>


I have also renamed the kspDeps as suggested from
https://github.com/com-lihaoyi/mill/pull/5727

---------

Co-authored-by: Tobias Roeser <le.petit.fou@web.de>

lefou committed:
Update classgraph to 4.8.181 (#5754)

Pull request: https://github.com/com-lihaoyi/mill/pull/5754

lefou committed:
Update log4j2-core to 2.25.1 (#5756)

Pull request: https://github.com/com-lihaoyi/mill/pull/5756

lefou committed:
Update spring-boot-loader-tools to 3.5.5 (#5755)

Pull request: https://github.com/com-lihaoyi/mill/pull/5755

dependabot[bot] committed:
Bump actions/setup-java from 4 to 5 (#5748)

Pull request: https://github.com/com-lihaoyi/mill/pull/5748

dependabot[bot] committed:
Bump actions/checkout from 4 to 5 (#5690)

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to
5.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/actions/checkout/releases">actions/checkout's
releases</a>.</em></p>
<blockquote>
<h2>v5.0.0</h2>
<h2>What's Changed</h2>
<ul>
<li>Update actions checkout to use node 24 by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2226">actions/checkout#2226</a></li>
<li>Prepare v5.0.0 release by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2238">actions/checkout#2238</a></li>
</ul>
<h2>⚠️ Minimum Compatible Runner Version</h2>
<p><strong>v2.327.1</strong><br />
<a
href="https://github.com/actions/runner/releases/tag/v2.327.1">Release
Notes</a></p>
<p>Make sure your runner is updated to this version or newer to use this
release.</p>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4...v5.0.0">https://github.com/actions/checkout/compare/v4...v5.0.0</a></p>
<h2>v4.3.0</h2>
<h2>What's Changed</h2>
<ul>
<li>docs: update README.md by <a
href="https://github.com/motss"><code>@​motss</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1971">actions/checkout#1971</a></li>
<li>Add internal repos for checking out multiple repositories by <a
href="https://github.com/mouismail"><code>@​mouismail</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1977">actions/checkout#1977</a></li>
<li>Documentation update - add recommended permissions to Readme by <a
href="https://github.com/benwells"><code>@​benwells</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2043">actions/checkout#2043</a></li>
<li>Adjust positioning of user email note and permissions heading by <a
href="https://github.com/joshmgross"><code>@​joshmgross</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2044">actions/checkout#2044</a></li>
<li>Update README.md by <a
href="https://github.com/nebuk89"><code>@​nebuk89</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2194">actions/checkout#2194</a></li>
<li>Update CODEOWNERS for actions by <a
href="https://github.com/TingluoHuang"><code>@​TingluoHuang</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/2224">actions/checkout#2224</a></li>
<li>Update package dependencies by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2236">actions/checkout#2236</a></li>
<li>Prepare release v4.3.0 by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2237">actions/checkout#2237</a></li>
</ul>
<h2>New Contributors</h2>
<ul>
<li><a href="https://github.com/motss"><code>@​motss</code></a> made
their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/1971">actions/checkout#1971</a></li>
<li><a href="https://github.com/mouismail"><code>@​mouismail</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/1977">actions/checkout#1977</a></li>
<li><a href="https://github.com/benwells"><code>@​benwells</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/2043">actions/checkout#2043</a></li>
<li><a href="https://github.com/nebuk89"><code>@​nebuk89</code></a> made
their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/2194">actions/checkout#2194</a></li>
<li><a href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/2236">actions/checkout#2236</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4...v4.3.0">https://github.com/actions/checkout/compare/v4...v4.3.0</a></p>
<h2>v4.2.2</h2>
<h2>What's Changed</h2>
<ul>
<li><code>url-helper.ts</code> now leverages well-known environment
variables by <a href="https://github.com/jww3"><code>@​jww3</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/1941">actions/checkout#1941</a></li>
<li>Expand unit test coverage for <code>isGhes</code> by <a
href="https://github.com/jww3"><code>@​jww3</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1946">actions/checkout#1946</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4.2.1...v4.2.2">https://github.com/actions/checkout/compare/v4.2.1...v4.2.2</a></p>
<h2>v4.2.1</h2>
<h2>What's Changed</h2>
<ul>
<li>Check out other refs/* by commit if provided, fall back to ref by <a
href="https://github.com/orhantoy"><code>@​orhantoy</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1924">actions/checkout#1924</a></li>
</ul>
<h2>New Contributors</h2>
<ul>
<li><a href="https://github.com/Jcambass"><code>@​Jcambass</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/1919">actions/checkout#1919</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4.2.0...v4.2.1">https://github.com/actions/checkout/compare/v4.2.0...v4.2.1</a></p>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Changelog</summary>
<p><em>Sourced from <a
href="https://github.com/actions/checkout/blob/main/CHANGELOG.md">actions/checkout's
changelog</a>.</em></p>
<blockquote>
<h1>Changelog</h1>
<h2>V5.0.0</h2>
<ul>
<li>Update actions checkout to use node 24 by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2226">actions/checkout#2226</a></li>
</ul>
<h2>V4.3.0</h2>
<ul>
<li>docs: update README.md by <a
href="https://github.com/motss"><code>@​motss</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1971">actions/checkout#1971</a></li>
<li>Add internal repos for checking out multiple repositories by <a
href="https://github.com/mouismail"><code>@​mouismail</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1977">actions/checkout#1977</a></li>
<li>Documentation update - add recommended permissions to Readme by <a
href="https://github.com/benwells"><code>@​benwells</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2043">actions/checkout#2043</a></li>
<li>Adjust positioning of user email note and permissions heading by <a
href="https://github.com/joshmgross"><code>@​joshmgross</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2044">actions/checkout#2044</a></li>
<li>Update README.md by <a
href="https://github.com/nebuk89"><code>@​nebuk89</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2194">actions/checkout#2194</a></li>
<li>Update CODEOWNERS for actions by <a
href="https://github.com/TingluoHuang"><code>@​TingluoHuang</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/2224">actions/checkout#2224</a></li>
<li>Update package dependencies by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2236">actions/checkout#2236</a></li>
</ul>
<h2>v4.2.2</h2>
<ul>
<li><code>url-helper.ts</code> now leverages well-known environment
variables by <a href="https://github.com/jww3"><code>@​jww3</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/1941">actions/checkout#1941</a></li>
<li>Expand unit test coverage for <code>isGhes</code> by <a
href="https://github.com/jww3"><code>@​jww3</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1946">actions/checkout#1946</a></li>
</ul>
<h2>v4.2.1</h2>
<ul>
<li>Check out other refs/* by commit if provided, fall back to ref by <a
href="https://github.com/orhantoy"><code>@​orhantoy</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1924">actions/checkout#1924</a></li>
</ul>
<h2>v4.2.0</h2>
<ul>
<li>Add Ref and Commit outputs by <a
href="https://github.com/lucacome"><code>@​lucacome</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1180">actions/checkout#1180</a></li>
<li>Dependency updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a>- <a
href="https://redirect.github.com/actions/checkout/pull/1777">actions/checkout#1777</a>,
<a
href="https://redirect.github.com/actions/checkout/pull/1872">actions/checkout#1872</a></li>
</ul>
<h2>v4.1.7</h2>
<ul>
<li>Bump the minor-npm-dependencies group across 1 directory with 4
updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1739">actions/checkout#1739</a></li>
<li>Bump actions/checkout from 3 to 4 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1697">actions/checkout#1697</a></li>
<li>Check out other refs/* by commit by <a
href="https://github.com/orhantoy"><code>@​orhantoy</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1774">actions/checkout#1774</a></li>
<li>Pin actions/checkout's own workflows to a known, good, stable
version. by <a href="https://github.com/jww3"><code>@​jww3</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1776">actions/checkout#1776</a></li>
</ul>
<h2>v4.1.6</h2>
<ul>
<li>Check platform to set archive extension appropriately by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1732">actions/checkout#1732</a></li>
</ul>
<h2>v4.1.5</h2>
<ul>
<li>Update NPM dependencies by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1703">actions/checkout#1703</a></li>
<li>Bump github/codeql-action from 2 to 3 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1694">actions/checkout#1694</a></li>
<li>Bump actions/setup-node from 1 to 4 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1696">actions/checkout#1696</a></li>
<li>Bump actions/upload-artifact from 2 to 4 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1695">actions/checkout#1695</a></li>
<li>README: Suggest <code>user.email</code> to be
<code>41898282+github-actions[bot]@users.noreply.github.com</code> by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1707">actions/checkout#1707</a></li>
</ul>
<h2>v4.1.4</h2>
<ul>
<li>Disable <code>extensions.worktreeConfig</code> when disabling
<code>sparse-checkout</code> by <a
href="https://github.com/jww3"><code>@​jww3</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1692">actions/checkout#1692</a></li>
<li>Add dependabot config by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1688">actions/checkout#1688</a></li>
<li>Bump the minor-actions-dependencies group with 2 updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1693">actions/checkout#1693</a></li>
<li>Bump word-wrap from 1.2.3 to 1.2.5 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1643">actions/checkout#1643</a></li>
</ul>
<h2>v4.1.3</h2>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/actions/checkout/commit/08c6903cd8c0fde910a37f88322edcfb5dd907a8"><code>08c6903</code></a>
Prepare v5.0.0 release (<a
href="https://redirect.github.com/actions/checkout/issues/2238">#2238</a>)</li>
<li><a
href="https://github.com/actions/checkout/commit/9f265659d3bb64ab1440b03b12f4d47a24320917"><code>9f26565</code></a>
Update actions checkout to use node 24 (<a
href="https://redirect.github.com/actions/checkout/issues/2226">#2226</a>)</li>
<li>See full diff in <a
href="https://github.com/actions/checkout/compare/v4...v5">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=actions/checkout&package-manager=github_actions&previous-version=4&new-version=5)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Tobias Roeser <le.petit.fou@web.de>

lefou committed:
Rework Android SDK installation (#5766)

This refactors the task dependency flow of `AndroidSdkModule`.
Previously, most tasks provided `PathRef`s to tools depending itself on
some anonymous installation tasks which needed to synchronize their
work.

Now, the installation steps are bound to single tasks, which return a
handle to the installed resources. These tasks are:

* `cmdlineTools`
* `androidSdk`
* `androidNdk`

Existing tasks pointing to various resources and executables, now just
depend on these installation tasks and simply forward the returned
resources.

Additional on Windows, tools will be searched for with the extensions
`exe` and `bat`.

This is not the final design, but a step forward in making the setup
more idiomatic and readable.

Pull request: https://github.com/com-lihaoyi/mill/pull/5766

vaslabs committed:
Android: Fix androidtodo and replace compose samples in ci with it (#5770)

This was broken accidentally since [the ksp 2
work](https://github.com/com-lihaoyi/mill/pull/5727) , I didn't realise
it wasn't running in the CI.


I'd prefer to have androidtodo over jetnews, as jetnews doesn't use
Hilt, and androidtodo uses both compose and hilt (and KSP), so if I may
choose one I prefer androidtodo.

nafg committed:
Fix typo in GitHub Dependency Graph submission text (#5772)

Please open all PRs as drafts and ensure that your fork of Mill has 
`settings/actions` / `Allow all actions and reusable workflows` enabled
to run CI on
your own fork of the Mill repo. Only once CI passes mark the PR as
`Ready for review`
and CI will run on the main Mill repo before we merge it.

lefou committed:
Merge KSP 1 and KSP 2 implementations into one module (#5768)

Consolidate all tasks into `KspModule` and introduce a config switch
`kspModuleMode` (not a task) to have different task dependency depending
on it's value.

All tasks specific to only one version have a `ksp1` or `ksp2` in their
name.

All unit and integration tests should keep running successfully.

I also envision to support KSP 2 In-Process mode, hence I implemented
the mode switch as enum.

Fix https://github.com/com-lihaoyi/mill/issues/5765

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>

souvlakias committed:
Android: Bugfixes for Windows Support (#5775)

### Motivation
Windows support was largely absent due to several bugs and issues. This
PR extends some fixes of https://github.com/com-lihaoyi/mill/pull/5766
and resolves additional of issues, enabling the basics of Android
Development with Mill work on Windows.
### Changes / Bugfixes
- Handling of `.bat` and `.exe` extensions
- A bug in `cmdlineToolsURL` 
- A bug in `acceptLicenses` (piping with `echo y` alone did not work on
windows)
- Skip `posix permission fix` of `androidLibsClassesJarMetaInf` on
Windows
- Pass file arguments with `@files.txt` on CLI calls.
([Ref](https://github.com/com-lihaoyi/mill/issues/5762#issuecomment-3232786720))

### Example Tests Coverage
⚠️✅ : Works but some tests don't pass due to expecting unix paths or
commands.
- AndroidLib
	- Java 
	1,2,5,6 ✅ 
	3,4 ⚠️✅
	- Kotlin
	1,2 ✅
	3 ❌ Screenshot png does not get generated
	4 ⚠️✅
- Thirdparty
	- Androidtodo ⚠️✅
	- Android Compose✅ 
	- Android Endless Tunnel ✅ 

#### Android Todo PoW



https://github.com/user-attachments/assets/6bac61ca-35eb-48ad-813b-4e705d2dcca9


	
### Notes
The `@files.txt` solution is only applied in a few tasks for now. Will
do some more testing and see if its applicable to other tasks that pass
many files to a CLI call as well.

lihaoyi committed:
Make Mill work with `mill-build/src/` folder even if `mill-build/build.mill` does not exist (#5783)

Fixes part of https://github.com/com-lihaoyi/mill/issues/5780, at least
the Mill handling of the `mill-build/` folder

Adjusted the `example/extending/metabuild/5-meta-shared-sources/` test
to work. Previously these modifications would cause this test to fail,
now it passes

Also tested manually to verify the IntelliJ import no longer exhibits
the weird behavior

lihaoyi committed:
Rebootstrap (#5784)
